### ISSUE TITLE
Broken symlinks make dune master fail on packages that previously compiled fine

### ISSUE THREAD

**Issue by kit-ty-kate** [ocaml/dune#3950]
## Expected Behavior

I would expect a warning that the symlink escape the scope of the project, not a failure.

## Actual Behavior

It happens for about ~10 package all around opam-repository. For instance with `logtk.1.6` (latest version)
```
#=== ERROR while compiling logtk.1.6 ==========================================#
# context           2.1.0~beta3 | linux/x86_64 | ocaml-base-compiler.4.11.1 | file:///home/opam/opam-repository
# path              ~/.opam/4.11.1/.opam-switch/build/logtk.1.6
# command           ~/.opam/4.11.1/bin/dune build -p logtk -j 47
# exit-code         1
# env-file          ~/.opam/log/logtk-7-4b0d17.env
# output-file       ~/.opam/log/logtk-7-4b0d17.out
### output ###
# Error: stat: examples/DAT013=1.p: No such file or directory
```
This happens because the symlink points to a file outside of the scope of the project and does not exist (cc @c-cube)
```
$ ls -l examples/DAT013=1.p
lrwxrwxrwx [...] 'examples/DAT013=1.p' -> '/home/simon/workspace/tptp/tptp/Problems/DAT/DAT013=1.p'
```

## Reproduction

1. Install dune master (after 558fd6cdf1b17d90289c86085fbf7425f2636b2b)
2. Install logtk.1.6

Looking at the code, I suspect 558fd6cdf1b17d90289c86085fbf7425f2636b2b (https://github.com/ocaml/dune/pull/3675) changed the behaviour as previously the symlink would get silently deleted if `Unix.lstat` were to encounter an issue (any exception)

**Comment by c-cube** [comment]
Indeed, I also get errors when building locally. It's true that dangling symlinks are not very clean, but they should not result in an error imho.

**Comment by kit-ty-kate** [comment]
Here is the set of packages that fail with this error:
```
logtk.1.6
logtk.1.5.1
libzipperposition.1.6
libzipperposition.1.5.1
oplsr.2.0.0
oplsr.5.0.1
oplsr.8.0.1
pardi.1.0.0
pardi.1.0.1
pardi.2.0.2
pardi.2.0.3
pardi.2.0.4
pardi.3.1.1
pardi.3.2.0
smbc.0.6
smbc.0.6.1
zipperposition.1.5.1
zipperposition.1.6
zipperposition-tools.1.6
```

**Comment by aalekseyev** [comment]
I was able to install logtk.1.6, both with the current master, and with dune version 558fd6c, so I'm a bit at a loss. 
I can see exactly this error with dune version 558fd6c^, though. Are you sure the version you tested was after 558fd6c?

**Comment by kit-ty-kate** [comment]
Oh, cool. It seems to have fixed itself since then my original report.
I had tried it with e50303f5f (merge of db18a122c – master at the time – and my own fix for https://github.com/ocaml/dune/pull/3904). This branch contained 558fd6c so i just assumed this was the issue.

I just merged master again (4d8c83a22) and this time it works.
I'm really unsure how it stopped working halfway through and started working again. This is a bit puzzling to me.

In any case, thanks a lot! I guess I can close this issue, I see you added a testcase in 558fd6cdf1 so everything is perfect. Thanks again and sorry for the noise.
